File: ruby/aoc/lib/aoc/day04b.rb
module AOC
  class Day04b < Day
    class Card
      attr_reader :id
      attr_reader :winning_numbers
      attr_reader :selected_numbers
      attr_accessor :copies

      def initialize(raw)
        game_ref, numbers = raw.split(": ")
        @id = game_ref.split(" ").last.to_i
        winning_numbers, selected_numbers = numbers.split(" | ")
        @winning_numbers = winning_numbers.split(" ").map(&:to_i)
        @selected_numbers = selected_numbers.split(" ").map(&:to_i)
        @copies = 1
      end

      def matches
        @matches ||= (@winning_numbers & @selected_numbers).length
      end

      def score
        return 0 if matches == 0
        return 1 if matches == 1
        return 2 ** (matches - 1)
      end

      def process(cards)
        ((id + 1)..(matches + id)).each do |i|
          card = cards.find { |c| c.id == i }
          card.copies += @copies
        end
      end
    end

    def sum_copies
      cards.each { |card| card.process(cards) }
      cards.map(&:copies).sum
    end

    def cards
      @cards ||= File.readlines(@path, chomp: true).map do |line|
        Card.new(line)
      end
    end
  end
end

File: elixir/aoc/lib/aoc/day04b.ex
defmodule AOC.Day04b do
  use AOC

  def solution(path) do
    path
    |> read_lines()
    |> Enum.map(&String.trim/1)
    |> Enum.map(&String.split(&1, ": "))
    |> Enum.map(&parse_card/1)
    |> Map.new()
    |> process_cards()
    |> Enum.map(fn {_id, {_score, count}} -> count end)
    |> Enum.sum()
  end

  def process_cards(cards) do
    cards
    |> Map.keys()
    |> Enum.sort()
    |> Enum.reduce(cards, &process_card/2)
  end

  def process_card(id, cards) do
    updated = copy_cards(id, cards[id], cards)
    Map.merge(cards, updated)
  end

  def copy_cards(_, {0, _}, _cards), do: %{}

  def copy_cards(id, {score, count}, cards) do
    (id + 1)..(id + score)
    |> Enum.map(&copy_card(&1, count, cards))
    |> Map.new()
  end

  def copy_card(id, duplicates, cards) do
    {score, count} = cards[id]
    {id, {score, count + duplicates}}
  end

  def parse_card([card_ref, numbers]) do
    id = card_ref |> String.split(" ", trim: true) |> List.last()
    [winning, ticket] = numbers |> String.split(" | ")
    score = score_card([winning, ticket])
    {String.to_integer(id), {score, 1}}
  end

  def score_card([winning, ticket]) do
    winning_numbers = winning |> String.split(" ", trim: true)
    ticket_numbers = ticket |> String.split(" ", trim: true)

    ticket_numbers
    |> Enum.filter(&Enum.member?(winning_numbers, &1))
    |> Enum.count()
  end
end
